src/Driver/PDO/Statement.php

```diff
@@ -117,12 +133,109 @@ public function execute($params = null): ResultInterface
             );
         }
 
+        $resourceOffsets = $this->getResourceOffsets();
         try {
             $this->stmt->execute($params);
         } catch (PDOException $exception) {
             throw Exception::new($exception);
+        } finally {
+            if ($resourceOffsets !== null) {
+                $this->restoreResourceOffsets($resourceOffsets);
+            }
         }
 
         return new Result($this->stmt);
     }
+
+    /**
+     * Track a binary parameter reference at binding time. These
+     * are cached for later analysis by the getResourceOffsets.
+     *
+     * @param mixed $resource
+     */
+    private function trackParamResource($resource): void
+    {
+        if (! is_resource($resource)) {
+            return;
+        }
+
+        $resources   = &$this->paramResources;
+        $resources ??= [];
+        $resources[] = $resource;
+    }
+
+    /**
+     * Determine the offset that any resource parameters needs to be
+     * restored to after the statement is executed. Call immediately
+     * before execute (not during bindValue) to get the most accurate offset.
+     *
+     * @return int[]|null Return offsets to restore if needed.
+     */
+    private function getResourceOffsets()
+    {
+        $resources = &$this->paramResources;
+        if ($resources === null) {
+            return null;
+        }
+
+        $count           = count($resources);
+        $resourceOffsets = null;
+        $deref           = [];
+        for ($i = 0; $i < $count; ++$i) {
```

Comment:
We're iterating over `$this->paramResources` here, don't we? Why don't we use a `foreach` loop?

Reply:
I'm using the $i for fill_array and keeping a parallel array for the offsets. AFAIK the foreach iteration order is not a guaranteed part of the PHP language--but this will always have the same order for the two parallel arrays. It would be unusual to have more than a couple of these, so I don't think there is a perf diff either way.

Reply:
The $deref variable on $183 isn't used though, so that needs to go.

Reply:
The $resourceOffsets array is now sparse. The iterator index is used to index into the sparse array.